docs/data-model/zng.md

```diff
@@ -61,69 +56,54 @@ or otherwise rewritten to be merged in this fashion.
-any ZNG streaming API.  In this way, senders and receivers of ZNG can embed
-protocol directives as ZNG control payloads rather than defining additional
-encapsulating protocols.
+Each frame header includes a length field
+allowing an implementation to easily skip from frame to frame.
 
-> For example, the [Zed service](../../docs/lake/service-api.md) query endpoint
-> uses application-defined message `0xfe` to embed search and server stats in
-> the return stream of ZNG data, e.g., as a long-running search progresses on
-> the server.
+Each frame begins with its version bit and a code indicating
+whether it is compressed or not:
+
+|    code    |  len    | body            |
+|------------|---------|-----------------|
+| `00xxxxxx` | uvarint | uncompressed data |
+| `01xxxxxx` | uvarint | compressed data |
+
+Bit 7 of the frame code must be zero as it defines version 0
+of the ZNG stream format.  If a future version of ZNG ever
+arises, bit 7 of future ZNG frames will be 1 and that version
+will define the interpretation of such frames.  ZNG version 0
+readers must ignore and skip over such frames using the
+`len` field, which must survive future versions.
+Any future versions of ZNG must be able to integrate version 0 frames
+for backward compatibility.
 
-### 2.1.1 Typedefs
+A frame of uncompressed data has the form:
 
-Following a header byte of `0xf5-0xfb` is a "typedef".  A typedef binds
-the smallest integer type ID not in use to a new type.  As there are
-a total of 30 primitive type IDs, the Type IDs for typedefs
-begin at the value 30 and increase by one for each typedef. These bindings
-are scoped to the stream in which the typedef occurs.
+|    code    |  len    | body            |
+|------------|---------|-----------------|
+| `0000xxxx` | uvarint | types data   |
+| `0001xxxx` | uvarint | values data  |
+| `0010xxxx` | uvarint | control data |
 
-Type IDs for the "primitive types" need not be defined with typedefs and
-are predefined with the IDs shown in the [Primitive Types](#-primitive-types) table.
+The frame `code` defines a version number and frame type,
+along with 4-bits enhancing the computed length
+as described below.
 
-A typedef is encoded as a single byte indicating the complex type ID followed by
-the type encoding.  This creates a binding between the implied type ID
-(i.e., 30 plus the count of all previous typedefs in the stream) and the new
-type definition.
+Bit 6 is `0` for an uncompressed frame.
 
-The type ID is encoded as a `uvarint`, an encoding used throughout the ZNG format.
+Bits 5 and 4 indicate the frame type:
+values (`00`), typedefs (`01`), control (`10`),
+or end of stream (`11`).
+
+The length of the frame body exclusive of the code and frame length
+is encoded in part as a variable length unsigned integer referred
+to here as a `uvarint`.  The frame length is equal to the uvarint `len`
+times 8 plus the low 4-bit integer value of the `code`.
```

Comment:
I don't feel strongly about this, but I do favor simplifying the specification and implementation by reserving the four extra code bits rather than adding them to the frame length. I think it would be a reasonable move since the decrease in entropy seems negligible, particularly for large frames. But again, this isn't something I feel strongly about!

Reply:
What would we reserve them for?  The version bit controls any future meaning of those bits.

Reply:
For making the specification and implementation simpler, but that doesn't strike me as an issue worth pursuing. What's here is fine.